src/xml_parser_main_metadata_together.c

```diff
@@ -0,0 +1,371 @@
+                                        void *newpkgcb_data,
+                                        cr_XmlParserPkgCb pkgcb,
+                                        void *pkgcb_data,
+                                        cr_XmlParserWarningCb warningcb,
+                                        void *warningcb_data,
+                                        GError **err)
+{
+    int ret = CRE_OK;
+    CR_FILE *primary_f;
+    CR_FILE *filelists_f;
+    CR_FILE *other_f;
+    GError *tmp_err = NULL;
+
+    cr_CbData cbdata;
+    cbdata.in_progress_pkgs = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, NULL);
+    cbdata.newpkgcb = newpkgcb;
+    cbdata.newpkgcb_data = newpkgcb_data;
+    cbdata.pkgcb = pkgcb;
+    cbdata.pkgcb_data = pkgcb_data;
+
+    //TODO(amatej): make filelists or other optional
+    assert(primary_path);
+    assert(filelists_path);
+    assert(other_path);
+    assert(newpkgcb || pkgcb);
+    assert(!err || *err == NULL);
+
+    primary_f = cr_open(primary_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", primary_path);
+        return code;
+    }
+    filelists_f = cr_open(filelists_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", filelists_path);
+        return code;
+    }
+    other_f = cr_open(other_path, CR_CW_MODE_READ, CR_CW_AUTO_DETECT_COMPRESSION, &tmp_err);
+    if (tmp_err) {
+        int code = tmp_err->code;
+        g_propagate_prefixed_error(err, tmp_err, "Cannot open %s: ", other_path);
+        return code;
+    }
+
+    cr_ParserData *primary_pd;
+    //TODO(amatej): replace do_files 0 by check if we are parsing filelists..
+    //TODO(amatej): if newpkgcb return NULL we should skip the pkg...
+    primary_pd = new_primary_parser_data(newpkgcb_primary, &cbdata, primary_pkgcb, &cbdata, warningcb, warningcb_data, 0);
+    cr_ParserData *filelists_pd;
+    filelists_pd = new_filelists_parser_data(newpkgcb_general, &cbdata, filelists_pkgcb, &cbdata, warningcb, warningcb_data);
+    cr_ParserData *other_pd;
+    other_pd = new_other_parser_data(newpkgcb_general, &cbdata, other_pkgcb, &cbdata, warningcb, warningcb_data);
+
+    char primary_is_done = 0;
+    char filelists_is_done = 0;
+    char other_is_done = 0;
+    cbdata.primary_count = 0;
+    cbdata.filelists_count = 0;
+    cbdata.other_count = 0;
+    while (1) {
+        if (!primary_is_done) {
+            do {
+                primary_is_done = parse_next_section(primary_f, primary_path, primary_pd, err);
+                if (*err) {
+                    ret = (*err)->code;
+                    goto out;
+                }
+
+            //TODO(amatej): what about not providing filelists? well at least one of them should always be provided
```

Comment:
Unrelated question.  What is the use case for having a subset of the files listed in primary.xml?  Is the subset enough for reliable depsolving?  What tools (if any) use that information and for what purposes?

And how was the heuristic decided?  https://github.com/rpm-software-management/createrepo_c/blob/8eff6ed99f5fd0ba844cb8513963435caab5fd3c/src/misc.h#L105-L118

Reply:
Good questions I will look into this more.
Maybe @dmach could know?

I think its not enough for all depsolving, there are use cases where dnf needs to download the filelists.

Reply:
@dmach I know you've moved on, but do you happen to know the answer to this question?  Or who I could ask?

Reply:
I suggest that the code is related to the future requirement to not loading/downloading filelists. When some package requires a file it is mostly located in ets, bin, user/bin, therefore to add all files from this location will allow to not require filelist in future (in theory).